File: src/lib/events.ts
// Engine event names + typed subscribers. Names mirror moin-core's events.rs.
//
// Everything about downloads arrives over the engine's shared event stream (see
// engine.ts). The one exception is the quit prompt: that's the desktop shell
// asking its own window a question, and never leaves the process.

import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { onEngineEvent } from "./engine";
import type {
  Category,
  Settings,
  Task,
  TaskProgress,
  ToolProgress,
} from "./types";

export const EV = {
  snapshot: "moin-task-snapshot",
  added: "moin-task-added",
  progressBatch: "moin-task-progress-batch",
  updated: "moin-task-updated",
  removed: "moin-task-removed",
  completed: "moin-task-completed",
  toolProgress: "moin-tool-progress",
  settingsChanged: "moin-settings-changed",
  categoriesChanged: "moin-categories-changed",
  confirmQuit: "moin-confirm-quit",
} as const;

/** Subscribe to the "confirm quit" prompt request (close attempted with active
 *  transfers and tray-minimize off); returns an unlisten function. A shell event,
 *  not an engine one. */
export function subscribeConfirmQuit(onConfirm: () => void): Promise<UnlistenFn> {
  return listen(EV.confirmQuit, () => onConfirm());
}

/** Subscribe to just the "a download was added" event — fires only on genuine
 *  new adds, not on the initial snapshot. Returns an unlisten function. */
export function subscribeTaskAdded(onAdded: () => void): Promise<UnlistenFn> {
  return Promise.resolve(onEngineEvent(EV.added, () => onAdded()));
}

export interface TaskHandlers {
  /** The full task list. Arrives first on every connection, and again if the
   *  stream ever had to resynchronise — so it replaces the set, never merges. */
  onSnapshot?: (tasks: Task[]) => void;
  onAdded?: (task: Task) => void;
  /** A coalesced batch of progress ticks (the engine flushes on a timer). */
  onProgress?: (batch: TaskProgress[]) => void;
  onUpdated?: (task: Task) => void;
  onRemoved?: (id: string) => void;
  /** A download finished — the client turns this into an OS notification. */
  onCompleted?: (task: Task) => void;
}

/** Subscribe to all task events; returns a single unlisten function. */
export async function subscribeTasks(h: TaskHandlers): Promise<UnlistenFn> {
  const off = [
    onEngineEvent<Task[]>(EV.snapshot, (tasks) => h.onSnapshot?.(tasks)),
    onEngineEvent<Task>(EV.added, (task) => h.onAdded?.(task)),
    onEngineEvent<TaskProgress[]>(EV.progressBatch, (b) => h.onProgress?.(b)),
    onEngineEvent<Task>(EV.updated, (task) => h.onUpdated?.(task)),
    onEngineEvent<string>(EV.removed, (id) => h.onRemoved?.(id)),
    onEngineEvent<Task>(EV.completed, (task) => h.onCompleted?.(task)),
  ];
  return () => off.forEach((fn) => fn());
}

/** Subscribe to aria2c binary-download progress; returns an unlisten function. */
export function subscribeToolProgress(
  onProgress: (p: ToolProgress) => void,
): Promise<UnlistenFn> {
  return Promise.resolve(onEngineEvent<ToolProgress>(EV.toolProgress, onProgress));
}

/** Settings changed — possibly in another window. Every view showing settings
 *  applies this so two windows can't drift apart. */
export function subscribeSettings(onChange: (s: Settings) => void): () => void {
  return onEngineEvent<Settings>(EV.settingsChanged, onChange);
}

/** The category list changed, from any window. */
export function subscribeCategories(
  onChange: (categories: Category[]) => void,
): () => void {
  return onEngineEvent<Category[]>(EV.categoriesChanged, onChange);
}

File: src-tauri/crates/moin-core/src/events.rs
//! Event names emitted to the frontend. Kept in one place so the TS side
//! (`src/lib/events.ts`) can mirror them exactly.
//!
//! Every source — HTTP now, torrent and media later — reports through this one
//! set, so the UI treats all downloads the same. Torrent-only fields (peers,
//! seeds, ratio) will ride along as optionals on the task payload.

/// The full task list, sent as the FIRST message on a fresh event subscription.
/// Subscribing and snapshotting in one stream is what closes the gap a separate
/// "fetch the list, then subscribe" would leave — no event can slip through
/// between the two. Also re-sent if a client ever falls far enough behind that
/// the broadcast dropped messages for it. Payload: `Vec<Task>`.
pub const TASK_SNAPSHOT: &str = "moin-task-snapshot";
/// A task was added to the queue. Payload: `Task`.
pub const TASK_ADDED: &str = "moin-task-added";
/// Frequent progress tick for an active task. Payload: `TaskProgress`.
pub const TASK_PROGRESS: &str = "moin-task-progress";
/// A whole interval's worth of progress ticks, coalesced into one event so the
/// IPC volume stays flat no matter how many torrents are active. Payload:
/// `Vec<TaskProgress>` (the latest reading per task since the last flush).
pub const TASK_PROGRESS_BATCH: &str = "moin-task-progress-batch";
/// A task's status changed (connecting, paused, done, failed…). Payload: `Task`.
pub const TASK_UPDATED: &str = "moin-task-updated";
/// A task was removed from the registry. Payload: the task id string.
pub const TASK_REMOVED: &str = "moin-task-removed";

/// A download just finished. The client turns this into an OS notification; the
/// engine only fires it when the user has notifications switched on. Payload: `Task`.
pub const TASK_COMPLETED: &str = "moin-task-completed";

/// Progress while a managed tool (aria2c) downloads its binary. Payload:
/// `ToolProgress` — received/total bytes of the archive.
pub const TOOL_PROGRESS: &str = "moin-tool-progress";

/// Settings changed, possibly from a different client. Every client applies this
/// so two open windows can't drift apart or overwrite each other's saves.
/// Payload: `Settings`.
pub const SETTINGS_CHANGED: &str = "moin-settings-changed";
/// The category list changed, from any client. Payload: `Vec<Category>`.
pub const CATEGORIES_CHANGED: &str = "moin-categories-changed";

/// The user tried to close the window while transfers are active and the tray
/// setting is off — the UI shows a confirm prompt (minimize / quit / cancel)
/// instead of quitting outright. No payload.
pub const CONFIRM_QUIT: &str = "moin-confirm-quit";
